src/main/scala/com/avast/marathon/plugin/vault/VaultPlugin.scala

```diff
@@ -12,47 +12,39 @@ import play.api.libs.json.{JsObject, _}
 
 import scala.util.{Failure, Success, Try}
 
-case class Configuration(address: String, token: String)
+case class Configuration(address: String, token: String, pathProvider: PathProvider)
+case class PathProvider(name: String, pathPrefix: Option[String])
 
 class VaultPlugin extends RunSpecTaskProcessor with PluginConfiguration {
 
   private val logger = LoggerFactory.getLogger(classOf[VaultPlugin])
   logger.info("Vault plugin instantiated")
 
   private var vault: Vault = _
+  private var pathProvider: VaultPathProvider = _
 
   override def initialize(marathonInfo: Map[String, Any], configurationJson: JsObject): Unit = {
-    val conf = configurationJson.as[Configuration](Json.reads[Configuration])
+    implicit val pathProviderFormat = Json.format[PathProvider]
+    val conf = configurationJson.as[Configuration](Json.format[Configuration])
     assert(conf != null, "VaultPlugin not initialized with configuration info.")
     assert(conf.address != null, "Vault address not specified.")
     assert(conf.token != null, "Vault token not specified.")
     vault = new Vault(new VaultConfig().address(conf.address).token(conf.token).build())
+    pathProvider = conf.pathProvider.name.toLowerCase match {
+      case "absolutepathprovider" => new AbsolutePathProvider()
+      case "relativepathprovider" => new RelativePathProvider(conf.pathProvider.pathPrefix.getOrElse(""))
+      case p => throw new RuntimeException(s"Unknown Vault path provider '$p'")
+    }
     logger.info(s"VaultPlugin initialized with $conf")
   }
 
-  /*
-    {
-      "env": {
-        "abc": "def",
-        "DB_PASSWORD": {
-          "secret": "db_pwd"
-        }
-      },
-      "secrets": {
-        "db_pwd": {
-          "source": "/path/to/vault/secret@password"
-        }
-      }
-    }
-  */
-
   def taskInfo(appSpec: ApplicationSpec, builder: TaskInfo.Builder): Unit = {
     val envBuilder = builder.getCommand.getEnvironment.toBuilder
     appSpec.env.foreach {
       case (name, v: EnvVarSecretRef) =>
         appSpec.secrets.get(v.secret) match {
           case Some(secret) =>
-            getSecretValueFromVault(secret) match {
+            getSecretValueFromVault(secret, pathProvider.getPath(appSpec, builder)) match {
```

Comment:
Why the `getPath` returns a function? So why cannot we just return a string from the method?

Reply:
I need to have `path` from secret

Reply:
Ok. The `getSecretValueFromVault` method is instance method so you can access the `pathProvider` there directly. So I would prefer to remove the second parameter of `getSecretValueFromVault` method, and call `pathProvider.getFullPath` from it.

Reply:
I don't understand how do you meant it. I need to pass `appSpec` and `builder` parameters (from the `taskInfo` method) and also `path` from the secret which is parsed in the `getSecretValueFromVault` method.